### PR TITLE
[CSBindings] Fix inference of partial key path from conversion constraints

### PR DESCRIPTION
**Pull request by xedin**
It's not permitted to use `PartialKeyPath` type to resolve a key path
expression, because that type is intended to be a type-erased version of
a fully resolved `KeyPath` type.

In situations where contextual type is a partial key path (e.g. parameter
type is a partial key path), let's replace it with a `KeyPath` type that
is a subtype of `PartialKeyPath` and allow value inference to happen.

Resolves: SR-5667
Resolves: rdar://32365183

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by jckarter — APPROVED** [review]
Thanks for fixing this!

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by xedin** [comment]
pretty please 🤞

**Comment by xedin** [comment]
@shahmishal Debug failed with `rm: /Users/buildnode/jenkins/workspace/swift-PR-source-compat-suite-debug/swift-source-compat-suite/project_cache/Kickstarter-ReactiveExtensions: Directory not empty`

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -356,6 +356,10 @@ class TypeVariableType::Implementation {
   /// Determine whether this type variable represents a closure result type.
   bool isClosureResultType() const;
 
+  /// Determine whether this type variable represetns
```
Comment (LucianoPAlmeida) [inline include/swift/Sema/ConstraintSystem.h:359]: Typo: represetns -> represents :)

Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h:359]: Let's see if this run of macOS succeeds if not - I'm going to fix it here...

Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h]: Fixed, thank you!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Pull request merged**